hypothesis-python/src/hypothesis/internal/conjecture/utils.py

```diff
@@ -172,14 +172,28 @@ def sample(self, data: "ConjectureData", forced: Optional[int] = None) -> int:
         forced_choice = (  # pragma: no branch # https://github.com/nedbat/coveragepy/issues/1617
             None
             if forced is None
-            else next((b, a, a_c) for (b, a, a_c) in self.table if forced in (b, a))
+            else next(
+                (b, a, a_c)
+                for (b, a, a_c) in self.table
+                if forced == b or (forced == a and a_c > 0)
+            )
         )
         base, alternate, alternate_chance = data.choice(
             self.table, forced=forced_choice
         )
-        use_alternate = data.draw_boolean(
-            alternate_chance, forced=None if forced is None else forced == alternate
-        )
+        forced_bool = None
+        if forced is not None:
+            # we maintain this invariant when picking forced_choice above.
+            # This song and dance about alternate_chance > 0 is to avoid forcing
+            # e.g. draw_boolean(p=0, forced=True), which is an error.
+            assert forced == base or (forced == alternate and alternate_chance > 0)
+            forced_bool = forced == alternate and alternate_chance > 0
+            # if alternate_chance == 0 (so the above conditional fails), we must
+            # have been equal to base.
+            if not forced_bool:
+                assert forced == base
```

Comment:
```suggestion
            forced_use_alternate = forced == alternate and alternate_chance > 0
            assert forced == base or forced_use_alternate
```
Renaming the variable makes this clear to me without the repetition.

Reply:
Thanks, this is clearer to me as well! I was struggling to find a clean explanation and ended up duplicating asserts, but wasn't particularly happy with it.